project-clone/internal/zip/setup.go

```diff
@@ -0,0 +1,168 @@
+		// Assume project is already set up
+		return nil
+	} else if err != nil {
+		return fmt.Errorf("failed to check path %s: %s", projectPath, err)
+	}
+
+	zipFilePath := path.Join(tmpDir, fmt.Sprintf("%s.zip", clonePath))
+	log.Printf("Downloading project archive from %s", url)
+	err := downloadZip(url, zipFilePath)
+	if err != nil {
+		return fmt.Errorf("failed to download archive: %s", err)
+	}
+
+	log.Printf("Extracting project archive to %s", zipFilePath)
+	err = unzip(zipFilePath, projectPath)
+	if err != nil {
+		return fmt.Errorf("failed to extract project zip archive: %s", err)
+	}
+
+	return nil
+}
+
+// downloadZip downloads file from `url` to `destPath`
+//
+// Adapted from the Che plugin broker:
+// https://github.com/eclipse/che-plugin-broker/blob/27e7c6953c92633cbe7e8ce746a16ca10d240ea2/utils/ioutil.go#L67
+func downloadZip(url, destPath string) error {
+	client := http.DefaultClient
+	resp, err := client.Get(url)
+	if err != nil {
+		return err
+	}
+	defer closeSafe(resp.Body)
+
+	if resp.StatusCode != http.StatusOK {
+		return fmt.Errorf("request at %s returned status code %d", url, resp.StatusCode)
+	}
+
+	out, err := os.Create(destPath)
+	if err != nil {
+		return err
+	}
+	defer closeSafe(out)
+
+	_, err = io.Copy(out, resp.Body)
+	if err != nil {
+		return err
+	}
+
+	return out.Sync()
+}
+
+// unzip extracts an archive to a destination path.
+//
+// Adapted from the Che plugin broker:
+// https://github.com/eclipse/che-plugin-broker/blob/27e7c6953c92633cbe7e8ce746a16ca10d240ea2/utils/ioutil.go#L190
+func unzip(archivePath string, destPath string) error {
+	r, err := zip.OpenReader(archivePath)
+	if err != nil {
+		return err
+	}
+	defer closeSafe(r)
+
+	if err := os.MkdirAll(destPath, 0755); err != nil {
+		return err
+	}
+
+	// Closure to address file descriptors issue with all the deferred .Close() methods
+	extractAndWriteFile := func(f *zip.File) error {
+		rc, err := f.Open()
+		if err != nil {
+			return err
+		}
+		defer func() {
+			if err := rc.Close(); err != nil {
+				panic(err)
+			}
+		}()
+
+		extractPath := filepath.Join(destPath, f.Name)
+
+		if f.FileInfo().IsDir() {
+			return os.MkdirAll(extractPath, 0755)
+		} else {
+			if err := os.MkdirAll(filepath.Dir(extractPath), 0775); err != nil {
+				return err
+			}
+			f, err := os.OpenFile(extractPath, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, f.Mode())
+			if err != nil {
+				return err
+			}
+			defer func() {
+				if err := f.Close(); err != nil {
+					panic(err)
+				}
+			}()
+
+			_, err = io.Copy(f, rc)
+			if err != nil {
+				return err
+			}
+
+			return f.Sync()
+		}
+	}
+
+	for _, f := range r.File {
+		err := extractAndWriteFile(f)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// closeSafe is a wrapper on io.Closer.Close() that just prints an error if encountered.
+//
+// Adapted from the Che plugin broker:
+// https://github.com/eclipse/che-plugin-broker/blob/27e7c6953c92633cbe7e8ce746a16ca10d240ea2/utils/ioutil.go#L326
```

Comment:
how these references from sources could help us in future? Maybe just drop them?

Reply:
It feels strange to copy-paste code from another project without referencing its source, even if I helped write both functions :)